### ISSUE TITLE
Custom gfx layers/rendering backends

### ISSUE THREAD

**Issue by bugproof** [c-smile/sciter-sdk#217]
Wanted to try sciter-lite in a game using vulkan but this backend is not supported

https://github.com/c-smile/sciter-sdk/blob/264ea65254d289fa62479d5c705ed73c07d12370/include/sciter-x-types.h#L32-L40

Is there any api that returns vertices and other stuff to draw? 

Currently I'm using RmlUi and it's very good you can implement your own RenderInterface SystemInterface and even font engine 

One option would be to create an overlay window and do the rendering there but this way it wouldn't be possible to integrate it more tightly with the game's rendering system e.g. draw 3d models on top of sciter UI

**Comment by pravic** [comment]
https://github.com/c-smile/sciter-js-sdk/tree/next Sciter.JS supports Vulkan. It's an early build (hence an alternative branch), but eventually it will come: https://sciter.com/vulkan-metal-dx12/, https://sciter.com/sciter-on-linux-with-vulkan/.

**Comment by pravic** [comment]
You don't need Sciter.Lite for that.

> Like I would like to add element for rendering models 

Where does your Vulkan requirement come from? Is your entire app in Vulkan or just your model renderer?

Create a regular Sciter app, add UI and everything. Implement your renderer using Vulkan and connect it via CSS via something like 

```css
modelrenderer { display: block; size: *; behavior: your-vulkan-renderer; }
```
Render your stuff using Vulkan and pass it to Sciter as a DirectX texture (I am assuming you're on Windows). Also Sciter UI can be rendered on top of (or in the background of) a 3D scene, so again - you don't need Sciter.Lite for that.

Sciter.Lite is used when the entire app is written without windows/widgets thus doesn't have or doesn't use standard keyboard/mouse input. Or [other extreme cases](https://sciter.com/windowless-sciter-android-and-iot-devices/) when you need to render your UI as a single bitmap and pass it somewhere to display.

**Comment by pravic** [comment]
> Also, is it possible to implement custom file interface so it could read custom texture formats from game archives without unpacking the files?

Of course, see the corresponding notifications: https://github.com/c-smile/sciter-sdk/blob/9f1724a45f5a53c4d513b02ed01cdbdab08fa0e5/include/sciter-x-host-callback.h#L129

**Comment by bugproof** [comment]
> You don't need Sciter.Lite for that.
> Where does your Vulkan requirement come from? Is your entire app in Vulkan or just your model renderer?

The game uses vulkan for rendering everything. I actually need Sciter.Lite because I need windowless sciter, I'm using SDL2 for all window related stuff.

> 
> Create a regular Sciter app, add UI and everything. Implement your renderer using Vulkan and connect it via CSS via something like
> 
> ```css
> modelrenderer { display: block; size: *; behavior: your-vulkan-renderer; }
> ```
> 

It's not very clear to me. What behavior actually is and how can I implement a behaviour that draws my 3d models using vulkan instance? It's not very obvious.

> Render your stuff using Vulkan and pass it to Sciter as a DirectX texture (I am assuming you're on Windows). Also Sciter UI can be rendered on top of (or in the background of) a 3D scene, so again - you don't need Sciter.Lite for that.

I want it to be rendered on top of a 3D scene but I also want a control over the rendering process so I can also render 3d stuff on top of Sciter. For example let's say I wanted to implement an inventory and display actual 3d models of items in the inventory.

**Comment by pravic** [comment]
> I actually need Sciter.Lite because I need windowless sciter, I'm using SDL2 for all window related stuff.

Okay, fair enough.

> What behavior actually is and how can I implement a behaviour that draws my 3d models using vulkan instance?

For example, especially in the end of the file: https://github.com/c-smile/sciter-sdk/blob/master/include/behaviors/behavior_drawing-opengl.cpp

**Comment by c-smile** [comment]
> I want it to be rendered on top of a 3D scene but I also want a control over the rendering process so I can also render 3d stuff on top of Sciter. For example let's say I wanted to implement an inventory and display actual 3d models of items in the inventory. It seems like Sciter doesn't give you any option to control the rendering process.

You can do that by using layered painting.

Let's assume that your document looks like this:
```html
<style>
   body { flow:stack; }
   body > section { size:*; /*spans whole body*/ }
</style>
<body >
    <section id="back-layer"> ... </section>
    <!-- your 3D stuff will go visually  here -->
    <section id="fore-layer"> ... </section>
</body >
 ```    
Note: we have there two layer elements - back and fore.

Then in order to render such sandwich we can use following drawing code:

```C++
{
      // draw background layer if needed
     SCITER_X_MSG_PAINT pc(back_layer, FALSE);
     SciterProcX(handle, pc);
 }

  draw_3d();

 {
    // draw foreground layer on top of 3d
     SCITER_X_MSG_PAINT pc(fore_layer, TRUE);
    SciterProcX(hWnd, pc);
}
```
where back/fore_layer  are  sciter::dom::element's set to  ```section id="back-layer"``` and ```section id="fore-layer"```
